interface/modules/custom_modules/oe-module-ehi-exporter/src/EhiExporter.php

```diff
@@ -0,0 +1,671 @@
+         */
+        while ($exportState->hasTableDefinitions() && $iterations++ <= $maxCycleLimit) {
+            $tableDefinition = $exportState->getNextTableDefinitionToProcess();
+            if ($this->shouldSkipTableDefinition($tableDefinition)) {
+                continue;
+            }
+            // otherwise if we have no records we skip as well.
+            $records = $tableDefinition->getRecords();
+            if (empty($records)) {
+                continue;
+            }
+
+            $keyDefinitions = $exportState->getKeyDataForTable($tableDefinition);
+            // write out the csv file
+            $this->writeCsvFile($jobTask, $records, $tableDefinition->table, $exportState->getTempSysDir());
+            $exportState->addExportResultTable($tableDefinition->table, count($records));
+            $tableDefinition->setHasNewData(false);
+            if (!empty($keyDefinitions)) {
+                foreach ($keyDefinitions['keys'] as $keyDefinition) {
+                    if (!($keyDefinition instanceof ExportKeyDefinition)) {
+                        throw new \RuntimeException("Invalid key definition");
+                    }
+                    $foreignKeyTableDefinition = $keyDefinitions['tables'][$keyDefinition->foreignKeyTable];
+                    // we process ALL parent keys, or if it is a child key we only process a select few of these keys.
+                    if ($this->shouldProcessForeignKey($keyDefinition)) {
+                        foreach ($records as $record) {
+                            $keyColumnName = $keyDefinition->localColumn;
+                            // we have in some cases a need to override the local value such as with our list_options
+                            // table so we can handle some more dynamic values here.
+                            if (isset($keyDefinition->localValueOverride)) {
+                                $recordValue = $keyDefinition->localValueOverride;
+                            } else {
+                                $recordValue = $record[$keyColumnName] ?? null;
+                            }
+                            if (isset($recordValue)) {
+                                $foreignKeyTableDefinition->addKeyValue($keyDefinition, $recordValue);
+                            }
+                        }
+                        // we only add it to be processed if there is new data to do so.
+                        if ($foreignKeyTableDefinition->hasNewData()) {
+                            // if the table already is in the queue the operation is a noop
+                            $exportState->addTableDefinition($foreignKeyTableDefinition);
+                        }
+                    }
+                }
+            }
+        }
+        $this->exportCustomTables($exportState);
+        if ($iterations > $maxCycleLimit) {
+            throw new \RuntimeException("Max iterations reached, check for cyclic dependencies");
+        }
```

Comment:
Its an arbitrary value to avoid infinite loops, I didn't think we'd ever hit 500 cycles.  We can raise the value but it makes me wonder if there is a bug in the key relationships or the code.  It loops through all of the tables but if it finds new data it will loop again and again until there is no further data found.  Is this happening on your large clinic staging server?

Reply:
I increased to 1000 and reduced fails by 3/4. I'll isolate some of the tables to see what is special with them. One I know is insurance numbers.

Reply:
> Is this happening on your large clinic staging server? yes

Finally completed at 1500 max. Took very long time though.